File: NativeScript/Source/Bindings/ECSBindings.cpp
#if UNITY_NATIVE_DLL

#include "Universal.h"
#include "Game.h"

#include "Helpers/SpawnHelpers.h"

#include "Context/GameStates.h"


/**
 * Bindings listing for C# layers to call, no C++ code should call this
 */


//----------------------------------------------------------------------------------------------------
// Template for auto-gen functions
#define DEFINE_GET_COMPONENT(ComponentType, FuncName)                                               \
ComponentType g_Static ##ComponentType;                                                             \
DLLEXPORT ComponentType FuncName(Entity e)	                                                        \
{																									\
    auto& registry = Game::Get()->GetCurrentWorld()->GetRegistry();                                 \
    if (registry.has<ComponentType>(e))                                                             \
	    return registry.get<ComponentType>(e);					                                    \
    else                                                                                            \
        return g_Static ##ComponentType;                                                            \
}																									\



#define DEFINE_GET_GAME_STATE(StateComponentType, FuncName)                             \
DLLEXPORT StateComponentType FuncName()                                                 \
{                                                                                       \
    auto& registry = Game::Get()->GetCurrentWorld()->GetRegistry();                     \
    const auto& gameStatesSingleton = registry.ctx<CtxGameStates>().holderEntity;       \
    return registry.get<StateComponentType>(gameStatesSingleton);                       \
}                                                                                       \



//----------------------------------------------------------------------------------------------------
// Spawn functions
DLLEXPORT void SpawnTower(float x, float y)
{
	SpawnHelpers::SpawnTower(Game::Get()->GetCurrentWorld()->GetRegistry(),
		x, y);
}



//----------------------------------------------------------------------------------------------------
// Auto generated section


/*BEGIN COMPONENT HEADER INCLUDE*/
#include "Components/BodySize.h"
#include "Components/CollisionLayer.h"
#include "Components/Cooldown.h"
#include "Components/CurrentWave.h"
#include "Components/Damage.h"
#include "Components/Destination.h"
#include "Components/ElapsedTime.h"
#include "Components/Health.h"
#include "Components/LifeTime.h"
#include "Components/MoveSpeed.h"
#include "Components/Orientation.h"
#include "Components/Position.h"
#include "Components/RangeDetection.h"
#include "Components/WaypointIndex.h"
/*END COMPONENT HEADER INCLUDE*/

/*BEGIN GET COMPONENT FUNCTIONS*/
DEFINE_GET_COMPONENT(CBodySize, GetBodySize)
DEFINE_GET_COMPONENT(CCollisionLayer, GetCollisionLayer)
DEFINE_GET_COMPONENT(CCooldown, GetCooldown)
DEFINE_GET_COMPONENT(CDamage, GetDamage)
DEFINE_GET_COMPONENT(CDestination, GetDestination)
DEFINE_GET_COMPONENT(CHealth, GetHealth)
DEFINE_GET_COMPONENT(CLifeTime, GetLifeTime)
DEFINE_GET_COMPONENT(CMoveSpeed, GetMoveSpeed)
DEFINE_GET_COMPONENT(COrientation, GetOrientation)
DEFINE_GET_COMPONENT(CPosition, GetPosition)
DEFINE_GET_COMPONENT(CRangeDetection, GetRangeDetection)
DEFINE_GET_COMPONENT(CWaypointIndex, GetWaypointIndex)
/*END GET COMPONENT FUNCTIONS*/


/*BEGIN GET STATES FUNCTIONS*/
DEFINE_GET_GAME_STATE(CCurrentWave, GetStateCurrentWave)
DEFINE_GET_GAME_STATE(CElapsedTime, GetStateElapsedTime)
/*END GET STATES FUNCTIONS*/

#endif

File: UnityProject/Assets/NativeScript/ECSBindings.cs
using NativeScript;
using System;
using System.Runtime.InteropServices;
using UnityEngine;

/// <summary>
/// Workaround for one-value components, which cannot be directly casted into
/// int or float values
/// </summary>
/// <typeparam name="T"></typeparam>
struct Value<T> 
{
    public T value;

    public Value(T value) { this.value = value; }

    public static implicit operator T(Value<T> a) => a.value;
    public static explicit operator Value<T>(T a) => new Value<T>(a);
}


/*BEGIN TYPE DEFINITION*/
public struct CBodySize
{
	public float radius;
}
public struct CCollisionLayer
{
	public System.Byte value;
}
public struct CCooldown
{
	public float remaining, max;
}
public struct CCurrentWave
{
	public int nextWaveIndex;
	public float lastSpawnInWave;
	public int numSpawnedInWave;
}
public struct CDamage
{
	public float value;
}
public struct CDestination
{
	public Vector3 value;
}
public struct CElapsedTime
{
	public float value;
}
public struct CHealth
{
	public float current;
	public float max;
}
public struct CLifeTime
{
	public float value;
}
public struct CMoveSpeed
{
	public float value;
}
public struct COrientation
{
	public Vector3 value;
}
public struct CPosition
{
	public Vector3 value;
}
public struct CRangeDetection
{
	public float range;
}
public struct CWaypointIndex
{
	public int value;
}
/*END TYPE DEFINITION*/


class ECSBindings
{
    /*BEGIN DELEGATE DECLARATION*/
	[UnmanagedFunctionPointer(CallingConvention.Cdecl)]
	public delegate CBodySize GetBodySizeType(UInt32 entity);
	public static GetBodySizeType GetBodySize;

	[UnmanagedFunctionPointer(CallingConvention.Cdecl)]
	public delegate CCollisionLayer GetCollisionLayerType(UInt32 entity);
	public static GetCollisionLayerType GetCollisionLayer;

	[UnmanagedFunctionPointer(CallingConvention.Cdecl)]
	public delegate CCooldown GetCooldownType(UInt32 entity);
	public static GetCooldownType GetCooldown;

	[UnmanagedFunctionPointer(CallingConvention.Cdecl)]
	public delegate CCurrentWave GetStateCurrentWaveType();
	public static GetStateCurrentWaveType GetStateCurrentWave;

	[UnmanagedFunctionPointer(CallingConvention.Cdecl)]
	public delegate CDamage GetDamageType(UInt32 entity);
	public static GetDamageType GetDamage;

	[UnmanagedFunctionPointer(CallingConvention.Cdecl)]
	public delegate CDestination GetDestinationType(UInt32 entity);
	public static GetDestinationType GetDestination;

	[UnmanagedFunctionPointer(CallingConvention.Cdecl)]
	public delegate CElapsedTime GetStateElapsedTimeType();
	public static GetStateElapsedTimeType GetStateElapsedTime;

	[UnmanagedFunctionPointer(CallingConvention.Cdecl)]
	public delegate CHealth GetHealthType(UInt32 entity);
	public static GetHealthType GetHealth;

	[UnmanagedFunctionPointer(CallingConvention.Cdecl)]
	public delegate CLifeTime GetLifeTimeType(UInt32 entity);
	public static GetLifeTimeType GetLifeTime;

	[UnmanagedFunctionPointer(CallingConvention.Cdecl)]
	public delegate CMoveSpeed GetMoveSpeedType(UInt32 entity);
	public static GetMoveSpeedType GetMoveSpeed;

	[UnmanagedFunctionPointer(CallingConvention.Cdecl)]
	public delegate COrientation GetOrientationType(UInt32 entity);
	public static GetOrientationType GetOrientation;

	[UnmanagedFunctionPointer(CallingConvention.Cdecl)]
	public delegate CPosition GetPositionType(UInt32 entity);
	public static GetPositionType GetPosition;

	[UnmanagedFunctionPointer(CallingConvention.Cdecl)]
	public delegate CRangeDetection GetRangeDetectionType(UInt32 entity);
	public static GetRangeDetectionType GetRangeDetection;

	[UnmanagedFunctionPointer(CallingConvention.Cdecl)]
	public delegate CWaypointIndex GetWaypointIndexType(UInt32 entity);
	public static GetWaypointIndexType GetWaypointIndex;

	/*END DELEGATE DECLARATION*/

    [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
    public delegate void SpawnTowerType(float x, float y);
    public static SpawnTowerType SpawnTower;

    public static void Load(IntPtr libraryHandle)
    {
        /*BEGIN DELEGATE ASSIGN*/
		GetBodySize = Bindings.GetDelegate<GetBodySizeType>(libraryHandle, "GetBodySize");
		GetCollisionLayer = Bindings.GetDelegate<GetCollisionLayerType>(libraryHandle, "GetCollisionLayer");
		GetCooldown = Bindings.GetDelegate<GetCooldownType>(libraryHandle, "GetCooldown");
		GetStateCurrentWave = Bindings.GetDelegate<GetStateCurrentWaveType>(libraryHandle, "GetStateCurrentWave");
		GetDamage = Bindings.GetDelegate<GetDamageType>(libraryHandle, "GetDamage");
		GetDestination = Bindings.GetDelegate<GetDestinationType>(libraryHandle, "GetDestination");
		GetStateElapsedTime = Bindings.GetDelegate<GetStateElapsedTimeType>(libraryHandle, "GetStateElapsedTime");
		GetHealth = Bindings.GetDelegate<GetHealthType>(libraryHandle, "GetHealth");
		GetLifeTime = Bindings.GetDelegate<GetLifeTimeType>(libraryHandle, "GetLifeTime");
		GetMoveSpeed = Bindings.GetDelegate<GetMoveSpeedType>(libraryHandle, "GetMoveSpeed");
		GetOrientation = Bindings.GetDelegate<GetOrientationType>(libraryHandle, "GetOrientation");
		GetPosition = Bindings.GetDelegate<GetPositionType>(libraryHandle, "GetPosition");
		GetRangeDetection = Bindings.GetDelegate<GetRangeDetectionType>(libraryHandle, "GetRangeDetection");
		GetWaypointIndex = Bindings.GetDelegate<GetWaypointIndexType>(libraryHandle, "GetWaypointIndex");
		/*END DELEGATE ASSIGN*/

        // Spawn functions
        SpawnTower = Bindings.GetDelegate<SpawnTowerType>(libraryHandle, "SpawnTower");
    }
}
